accounts-db/src/accounts_db.rs

```diff
@@ -2752,78 +2750,67 @@ impl AccountsDb {
             .expect("Cluster type must be set at initialization")
     }
 
-    /// Reclaim older states of accounts older than max_clean_root_inclusive for AccountsDb bloat mitigation.
-    /// Any accounts which are removed from the accounts index are returned in PubkeysRemovedFromAccountsIndex.
-    /// These should NOT be unref'd later from the accounts index.
-    fn clean_accounts_older_than_root(
+    /// While scanning cleaning candidates obtain slots that can be
+    /// reclaimed for each pubkey. In addition if the pubkey is
+    /// removed from the index, let the caller know this.
+    fn collect_reclaims(
         &self,
-        candidates: &[RwLock<HashMap<Pubkey, CleaningInfo>>],
+        pubkey: &Pubkey,
         max_clean_root_inclusive: Option<Slot>,
         ancient_account_cleans: &AtomicU64,
         epoch_schedule: &EpochSchedule,
-    ) -> (ReclaimResult, PubkeysRemovedFromAccountsIndex) {
-        let pubkeys_removed_from_accounts_index = HashSet::default();
+    ) -> (SlotList<AccountInfo>, bool) {
         let one_epoch_old = self.get_oldest_non_ancient_slot(epoch_schedule);
-        let pubkeys_removed_from_accounts_index = Mutex::new(pubkeys_removed_from_accounts_index);
-
         let mut clean_rooted = Measure::start("clean_old_root-ms");
-        let reclaim_vecs = candidates
-            .par_iter()
-            .filter_map(|candidates_bin| {
-                let mut reclaims = Vec::new();
-                for (pubkey, cleaning_info) in candidates_bin.read().unwrap().iter() {
-                    if cleaning_info.should_purge {
-                        let removed_from_index = self.accounts_index.clean_rooted_entries(
-                            pubkey,
-                            &mut reclaims,
-                            max_clean_root_inclusive,
-                        );
-                        if removed_from_index {
```

Comment:
I think if I were doing it, I'd keep this `pubkeys_removed_from_accounts_index` in this fn. You're already using the exact same data type: `Mutex<HashSet<...`. This is another case where we already have `HashMap<Pubkey, ...` and we could just re-use that hashmap. I don't think I love that. I doubt we are deleting that many keys per clean. I'm sure there's a metric for that. But, whatever the case, the main point is updating `pubkeys_removed_from_accounts_index` could easily remain where it is in this fn.

Reply:
With this change each invocation of the function operates on a single pubkey. I don't follow how creating a new HashMap for a single pubkey on each call to the function and returning it to the caller only to be moved into another HashMap is helpful. The HashMap that collects all pubkeys removed from accounts index still has to exist in the caller context, it seems. Maybe, it could be passed as a mutable parameter to this function -- is this what you're suggesting?

Reply:
no, pass `&Mutex<HashSet<...>>` or whatever so ALL parallel threads share this on each call.

Reply:
Done.